cloudflare_list_item fails and throws golang exception

### Confirmation

- [X] My issue isn't already found on the issue tracker.
- [X] I have replicated my issue using the latest version of the provider and it is still present.

### Terraform and Cloudflare provider version

~~~
❯ terraform -v
Terraform v1.6.4
on linux_amd64
+ provider registry.terraform.io/cloudflare/cloudflare v4.20.0
+ provider registry.terraform.io/hashicorp/aws v5.28.0
~~~

### Affected resource(s)

~~~
cloudflare_list_item
~~~

### Terraform configuration files

```hcl
resource "cloudflare_list" "bulk_redirect_to_id" {
  account_id  = var.cloudflare_account_id
  name        = var.bulk_redirect_list_name
  description = var.bulk_redirect_list_description
  kind        = "redirect"
}

resource "cloudflare_list_item" "bulk_redirect_to_id_item" {
   
  # the issue happens even with 1 resource
  for_each = { for redirect in var.bulk_redirects : "${redirect.source_url}" => redirect }

  account_id = var.cloudflare_account_id
  list_id    = cloudflare_list.bulk_redirect_to_id.id

  redirect {
    source_url  = each.value.source_url
    target_url  = each.value.target_url
    status_code = each.value.status_code
  }

  depends_on = [
    cloudflare_list.bulk_redirect_to_id
  ]

}

# Create Bulk Redirect and attach bulk redirect list
resource "cloudflare_ruleset" "bulk_root_redirect_to_id" {
  account_id  = var.cloudflare_account_id
  name        = var.bulk_redirect_list_name
  description = var.bulk_redirect_list_description
  kind        = "root"
  phase       = "http_request_redirect"

  rules {
    action = "redirect"
    action_parameters {
      from_list {
        name = var.bulk_redirect_list_name // name of bulk redirect list
        key  = "http.request.full_uri"
      }
    }
    expression  = "http.request.full_uri in ${"$"}${var.bulk_redirect_list_name}" // requires the $ sign
    description = var.bulk_redirect_list_description
    enabled     = true
  }

  depends_on = [
    cloudflare_list_item.bulk_redirect_to_id_item
  ]
}


## variables.tf
variable "cloudflare_account_id" {
  default = "123456"
}

variable "bulk_description" {
  default = "my desc"
}

variable "bulk_redirects" {
  type = map(object({
    source_url  = string
    target_url  = optional(string, "https://my.redir.example")
    status_code = optional(number, 301)
  }))

  default = {
    "demo"         = { source_url = "https://demo.example.com/" },
  }
}
```


### Link to debug output

https://gist.github.com/tchellomello/455227456ae0ad3fec86370ac87023a4

### Panic output

~~~
Stack trace from the terraform-provider-cloudflare_v4.20.0 plugin:

panic: interface conversion: interface {} is []interface {}, not string

goroutine 185 [running]:
github.com/cloudflare/terraform-provider-cloudflare/internal/sdkv2provider.getSearchTerm(0xc000500700?)
	github.com/cloudflare/terraform-provider-cloudflare/internal/sdkv2provider/resource_cloudflare_list_item.go:208 +0x1cd
github.com/cloudflare/terraform-provider-cloudflare/internal/sdkv2provider.resourceCloudflareListItemCreate({0x1397268, 0xc00091a180}, 0x0?, {0x1199e40?, 0xc000434a50})
	github.com/cloudflare/terraform-provider-cloudflare/internal/sdkv2provider/resource_cloudflare_list_item.go:61 +0x669
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).create(0xc00051f6c0, {0x13972a0, 0xc0008f40f0}, 0xd?, {0x1199e40, 0xc000434a50})
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.30.0/helper/schema/resource.go:778 +0x12e
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc00051f6c0, {0x13972a0, 0xc0008f40f0}, 0xc0006b2340, 0xc00055ed00, {0x1199e40, 0xc000434a50})
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.30.0/helper/schema/resource.go:909 +0xa7e
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc00074a018, {0x13972a0?, 0xc000937ef0?}, 0xc0003876d0)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.30.0/helper/schema/grpc_provider.go:1072 +0xe8d
github.com/hashicorp/terraform-plugin-mux/tf5to6server.v5tov6Server.ApplyResourceChange({{0x139daa0?, 0xc00074a018?}}, {0x13972a0, 0xc000937ef0}, 0x0?)
	github.com/hashicorp/terraform-plugin-mux@v0.12.0/tf5to6server/tf5to6server.go:37 +0x5a
github.com/hashicorp/terraform-plugin-mux/tf6muxserver.(*muxServer).ApplyResourceChange(0x13971f8?, {0x13972a0?, 0xc000937bc0?}, 0xc000387680)
	github.com/hashicorp/terraform-plugin-mux@v0.12.0/tf6muxserver/mux_server_ApplyResourceChange.go:36 +0x1b5
github.com/hashicorp/terraform-plugin-go/tfprotov6/tf6server.(*server).ApplyResourceChange(0xc00039f360, {0x13972a0?, 0xc0009373b0?}, 0xc000318150)
	github.com/hashicorp/terraform-plugin-go@v0.19.1/tfprotov6/tf6server/server.go:857 +0x574
github.com/hashicorp/terraform-plugin-go/tfprotov6/internal/tfplugin6._Provider_ApplyResourceChange_Handler({0x11660e0?, 0xc00039f360}, {0x13972a0, 0xc0009373b0}, 0xc00055ea00, 0x0)
	github.com/hashicorp/terraform-plugin-go@v0.19.1/tfprotov6/internal/tfplugin6/tfplugin6_grpc.pb.go:467 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc00078c000, {0x13972a0, 0xc0009372f0}, {0x139c258, 0xc000502d00}, 0xc00068d560, 0xc00078a8d0, 0x1a4d798, 0x0)
	google.golang.org/grpc@v1.59.0/server.go:1343 +0xe49
google.golang.org/grpc.(*Server).handleStream(0xc00078c000, {0x139c258, 0xc000502d00}, 0xc00068d560)
	google.golang.org/grpc@v1.59.0/server.go:1737 +0xca6
google.golang.org/grpc.(*Server).serveStreams.func1.1()
	google.golang.org/grpc@v1.59.0/server.go:986 +0x8c
created by google.golang.org/grpc.(*Server).serveStreams.func1
	google.golang.org/grpc@v1.59.0/server.go:997 +0x15c

Error: The terraform-provider-cloudflare_v4.20.0 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
~~~

### Expected output

Resources to be created

### Actual output

~~~
running "/usr/local/bin/terraform apply -input=false \"/xxxx/repos/xxxx/default.tfplan\"" in "/xxxx/repos/xxxxx": exit status 1
module.bulk_root_redirect_to_id.cloudflare_list.bulk_redirect_to_id: Creating...
module.bulk_root_redirect_to_id.cloudflare_ruleset.bulk_root_redirect_to_id: Creating...
module.bulk_root_redirect_to_id.cloudflare_list.bulk_redirect_to_id: Creation complete after 2s [id=123456]
module.bulk_root_redirect_to_id.cloudflare_list_item.bulk_redirect_to_id_item["https://demo1/"]: Creating...
module.bulk_root_redirect_to_id.cloudflare_list_item.bulk_redirect_to_id_item["https://demo2/"]: Creating...
module.bulk_root_redirect_to_id.cloudflare_list_item.bulk_redirect_to_id_item["https://demo3"]: Creating...
╷
│ Error: Plugin did not respond
│ 
│ The plugin encountered an error, and failed to respond to the
│ plugin6.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.
╵
╷
│ Error: Request cancelled
│ 
│ The plugin6.(*GRPCProvider).ValidateResourceConfig request was cancelled.
╵
╷
│ Error: Request cancelled
│ 
│ The plugin6.(*GRPCProvider).ValidateResourceConfig request was cancelled.
╵
╷
│ Error: Request cancelled
│ 
│ The plugin6.(*GRPCProvider).ValidateResourceConfig request was cancelled.
╵
╷
│ Error: Plugin did not respond
│ 
│ The plugin encountered an error, and failed to respond to the
│ plugin6.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.
╵
╷
│ Error: Request cancelled
│ 
│ The plugin6.(*GRPCProvider).ValidateResourceConfig request was cancelled.
╵
╷
│ Error: Plugin did not respond
│ 
│ The plugin encountered an error, and failed to respond to the
│ plugin6.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.
╵
╷
│ Error: Plugin did not respond
│ 
│ The plugin encountered an error, and failed to respond to the
│ plugin6.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.
╵
╷
│ Error: Request cancelled
│ 
│ The plugin6.(*GRPCProvider).ValidateResourceConfig request was cancelled.
╵
╷
│ Error: Plugin did not respond
│ 
│ The plugin encountered an error, and failed to respond to the
│ plugin6.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.
╵
╷
│ Error: Plugin did not respond
│ 
│ The plugin encountered an error, and failed to respond to the
│ plugin6.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.
╵
╷
│ Error: Request cancelled
│ 
│ The plugin6.(*GRPCProvider).ValidateResourceConfig request was cancelled.
╵
╷
│ Error: Plugin did not respond
│ 
│ The plugin encountered an error, and failed to respond to the
│ plugin6.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.
╵
╷
│ Error: Request cancelled
│ 
│ The plugin6.(*GRPCProvider).ValidateResourceConfig request was cancelled.
╵
╷
│ Error: Request cancelled
│ 
│ The plugin6.(*GRPCProvider).ValidateResourceConfig request was cancelled.
╵
╷
│ Error: Plugin did not respond
│ 
│ The plugin encountered an error, and failed to respond to the
│ plugin6.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.
╵
╷
│ Error: Request cancelled
│ 
│ The plugin6.(*GRPCProvider).ValidateResourceConfig request was cancelled.
╵
╷
│ Error: Plugin did not respond
│ 
│ The plugin encountered an error, and failed to respond to the
│ plugin6.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.
╵
╷
│ Error: Plugin did not respond
│ 
│ The plugin encountered an error, and failed to respond to the
│ plugin6.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.
╵
╷
│ Error: Request cancelled
│ 
│ The plugin6.(*GRPCProvider).ValidateResourceConfig request was cancelled.
╵
~~~

### Steps to reproduce

1. Invoke the module
2. Issue is reproduced

### Additional factoids

What is interesting is that the object is being created on the Cloudflare's side, but the TF statefile never gets object created. 

Looking at the Go trace, we can see its related with this part of the code: https://github.com/cloudflare/terraform-provider-cloudflare/blob/master/internal/sdkv2provider/resource_cloudflare_list_item.go#L59-L74

Troubleshooting it further, I found this commit https://github.com/cloudflare/terraform-provider-cloudflare/commit/ed28816da148cd1c7fa538ff4070d70fba20e8e2 which seems to be related with part of the problem. 

To test it, I've downgraded the provider to `v.4.18.0` and then I found a different error as shown below: 

~~~
❯ terraform apply
data.aws_secretsmanager_secret.secret: Reading...
data.aws_secretsmanager_secret.secret: Read complete after 0s [id=arn:aws:xxxxxxx]
data.aws_secretsmanager_secret_version.secret: Reading...
data.aws_secretsmanager_secret_version.secret: Read complete after 0s [id=arn:aws:xxx|AWSCURRENT]
module.bulk_root_redirect_to_id.cloudflare_list.bulk_redirect_to_id: Refreshing state... [id=12345]
module.bulk_root_redirect_to_id.cloudflare_list_item.bulk_redirect_to_id_item["https://demo1/"]: Refreshing state... [id=xxxxx]
module.bulk_root_redirect_to_id.cloudflare_list_item.bulk_redirect_to_id_item["https://demo2/"]: Refreshing state... [id=xxxyy]


Planning failed. Terraform encountered an error while generating this plan.

╷
│ Error: error reading List Item with ID "xxxxxxxxx": Authentication error (10000)
│ 
│   with module.bulk_root_redirect_to_id.cloudflare_list_item.bulk_redirect_to_id_item["demo4/"],
│   on ../../../modules/bulk_redirects/main.tf line 9, in resource "cloudflare_list_item" "bulk_redirect_to_id_item":
│    9: resource "cloudflare_list_item" "bulk_redirect_to_id_item" {
~~~

The main difference here is that with the version `v4.18.0`, the objects were created in the API and **also** in the TF state file.

So basically during the troubleshooting, downgrading to `v4.18.0`, we were able to create the bulk_list and then upgrading to `v4.20.0` (and with the statefile populated) we were able to link the list to a bulk name.




### References

_No response_

--- a/.changelog/2975.txt
+++ b/.changelog/2975.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+resource/cloudflare_list_item: fix issue preventing usage of redirect item type
+```

--- a/internal/sdkv2provider/resource_cloudflare_list_item.go
+++ b/internal/sdkv2provider/resource_cloudflare_list_item.go
@@ -137,14 +137,17 @@ func resourceCloudflareListItemRead(ctx context.Context, d *schema.ResourceData,
 			}
 			return ""
 		}
-
-		d.Set("source_url", listItem.Redirect.SourceUrl)
-		d.Set("include_subdomains", optBoolToString(listItem.Redirect.IncludeSubdomains))
-		d.Set("target_url", listItem.Redirect.TargetUrl)
-		d.Set("status_code", listItem.Redirect.StatusCode)
-		d.Set("preserve_query_string", optBoolToString(listItem.Redirect.PreserveQueryString))
-		d.Set("subpath_matching", optBoolToString(listItem.Redirect.SubpathMatching))
-		d.Set("preserve_path_suffix", optBoolToString(listItem.Redirect.PreservePathSuffix))
+		d.Set("redirect", []interface{}{
+			map[string]interface{}{
+				"source_url":            listItem.Redirect.SourceUrl,
+				"include_subdomains":    optBoolToString(listItem.Redirect.IncludeSubdomains),
+				"target_url":            listItem.Redirect.TargetUrl,
+				"status_code":           listItem.Redirect.StatusCode,
+				"preserve_query_string": optBoolToString(listItem.Redirect.PreserveQueryString),
+				"subpath_matching":      optBoolToString(listItem.Redirect.SubpathMatching),
+				"preserve_path_suffix":  optBoolToString(listItem.Redirect.PreservePathSuffix),
+			},
+		})
 	}
 
 	return nil
@@ -205,7 +208,7 @@ func getSearchTerm(d *schema.ResourceData) string {
 	}
 
 	if redirect, ok := d.GetOk("redirect"); ok {
-		return redirect.(string)
+		return redirect.([]interface{})[0].(map[string]interface{})["source_url"].(string)
 	}
 
 	if hostname, ok := d.GetOk("hostname"); ok {

--- a/internal/sdkv2provider/resource_cloudflare_list_item_test.go
+++ b/internal/sdkv2provider/resource_cloudflare_list_item_test.go
@@ -178,6 +178,38 @@ func TestAccCloudflareListItem_Hostname(t *testing.T) {
 	})
 }
 
+func TestAccCloudflareListItem_Redirect(t *testing.T) {
+	rnd := generateRandomResourceName()
+	name := fmt.Sprintf("cloudflare_list_item.%s", rnd)
+	accountID := os.Getenv("CLOUDFLARE_ACCOUNT_ID")
+
+	var ListItem cloudflare.ListItem
+
+	resource.Test(t, resource.TestCase{
+		PreCheck: func() {
+			testAccPreCheck(t)
+			testAccPreCheckAccount(t)
+		},
+		ProviderFactories: providerFactories,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccCheckCloudflareHostnameRedirectItem(rnd, rnd, rnd, accountID),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckCloudflareListItemExists(name, rnd, &ListItem),
+					resource.TestCheckResourceAttr(
+						name, "redirect.#", "1"),
+					resource.TestCheckResourceAttr(
+						name, "redirect.0.source_url", "example.com/"),
+					resource.TestCheckResourceAttr(
+						name, "redirect.0.target_url", "https://example1.com"),
+					resource.TestCheckResourceAttr(
+						name, "redirect.0.status_code", "301"),
+				),
+			},
+		},
+	})
+}
+
 func testAccCheckCloudflareIPListItem(ID, name, comment, accountID string) string {
 	return fmt.Sprintf(`
   resource "cloudflare_list" "%[2]s" {
@@ -308,6 +340,27 @@ func testAccCheckCloudflareHostnameListItem(ID, name, comment, accountID string)
 	comment    = "%[3]s"
 	hostname {
 		url_hostname = "example.com"
+	}
+  } `, ID, name, comment, accountID)
+}
+
+func testAccCheckCloudflareHostnameRedirectItem(ID, name, comment, accountID string) string {
+	return fmt.Sprintf(`
+  resource "cloudflare_list" "%[2]s" {
+    account_id          = "%[4]s"
+    name                = "%[2]s"
+    description         = "list named %[2]s"
+    kind                = "redirect"
+  }
+
+  resource "cloudflare_list_item" "%[1]s" {
+    account_id = "%[4]s"
+	list_id    = cloudflare_list.%[2]s.id
+	comment    = "%[3]s"
+	redirect {
+		source_url = "example.com/"
+		target_url = "https://example1.com"
+		status_code = 301
 	}
   } `, ID, name, comment, accountID)
 }

